Implements heat/orchestration stack update PATCH verb.

scott-chaney opened the pull request:
For #[[1101](https://github.com/gophercloud/gophercloud/issues/1101)]

Related Openstack code:
https://github.com/openstack/heat/blob/master/heat/api/openstack/v1/stacks.py#L494
https://github.com/openstack/heat/blob/master/heat/api/openstack/v1/__init__.py#L218

Some notes on the implementation:
Both PUT and PATCH update verbs are acceptable, but the PATCH verb allows template parameter updates without needing to specify the template to the API call, where as PUT does require the template.  PATCH does accept a template, but does not delete the environment overlay as PUT does.  See `openstack/orchestration/v1/stacks/doc.go` for more information on PUT vs PATCH functionality.

I ran this implementation by my colleagues and gophercloud contributors @problemv, @JackKuei  and @dkt26111 and we had some uncertainty on the most gophercloud appropriate manner to implement the PUT vs PATCH, specifically on both routes usage of` ToStackUpdateMaps()`.

We explored the following options:
- Current method of having the Update functions pass in an "existing" flag to the `ToStackUpdateMap() `to reason if a template is required
- Adding a private member to the `UpdateOpts` struct, "existing", that is set by the Update functions and then checked via new interface method in the `ToStackUpdateMap()` function
- Using completely divergent paths for `Update()` and `UpdatePatch()` functions (new structs, new interfaces, new functions)
- Moving the logic from the `ToStackUpdateMap()` function into the different update functions

Since this feature of distinguishing between PUT and PATCH is somewhat unique to stack updates in GopherCloud,  we are hoping to get some feedback on which implementation fits best with in the language and project paradigms.

scott-chaney commented:
recheck

scott-chaney commented:
The failure noted on the checks doesn't give much info, and from what I can tell the recheck didn't trigger the check again. Hoping someone who has access to the logs can take a look and let me know if it was a problem on the code side or a hiccup in the CI job. Thanks!

jtopjian commented:
@scott-chaney It looks like the failure is from Travis and I've gone ahead and restarted the job. The `recheck` command is only for OpenLab tests, confusingly.

jtopjian commented:
@scott-chaney Thank you (and the rest of the team) for looking into this. 

Given the minor difference between the `Update` and `UpdatePatch` functions, I think we can proceed with reusing `UpdateOpts` for both.

The modification to the interface method's signature (`existing`) works, but I'm a little hesitant to do that just yet.

What if we did something like this: https://github.com/jtopjian/gophercloud/commit/3dd3cc2b7888b47f1da21d1dd25df88cf458783b

The idea is that we extract the common code of building the map to a private function in the package. We then have two interfaces, both of which `UpdateOpts` can satisfy. One of the interface methods guards against a nil `TemplateOpts`. 

I quickly put this together so some comments might need changed, maybe a better name for the private function, etc. Or this might not be a good approach at all.

Thoughts?

scott-chaney commented:
Hey @jtopjian thanks for the feedback! This looks like a cleaner implementation.  I didn't like having to change the function signature and pass a random bool.  I'll integrate these changes and test them out.  Thanks!

jtopjian commented on openstack/orchestration/v1/stacks/doc.go:
```diff
@@ -5,4 +5,121 @@
 // application framework or component specified (in the template). A stack is a
 // running instance of a template. The result of creating a stack is a deployment
 // of the application framework or component.
+
```
More documentation is always better than no documentation, so we can definitely keep everything here. However, let's strip out all `//` and `/* */` marks from this document and just have one `/*` at the beginning and one `*/` at the end of this file.

jtopjian commented on openstack/orchestration/v1/stacks/doc.go:
```diff
@@ -5,4 +5,121 @@
+raw_template value.
+*/
+
+// example stack update using the Update() (PUT) method
```
Example to Update a Stack using the Update (PUT) Method

Lines like this should have a "title" format using upper-case when appropriate.

jtopjian commented on openstack/orchestration/v1/stacks/doc.go:
```diff
@@ -5,4 +5,121 @@
+	TemplateOpts: &template,
+}
+
+// client is an instance of *gophercloud.ServiceClient
```
Instead of this line, just call client `orchestrationClient` or something like that.

jtopjian commented on openstack/orchestration/v1/stacks/doc.go:
```diff
@@ -5,4 +5,121 @@
+//example template.yaml of a Heat::ResourceGroup with 3 nodes
+*/
+
+//example stack update using the UpdatePatch()  (PATCH) method
```
Example to Update a Stack using the UpdatePatch (PATCH) Method

jtopjian commented on openstack/orchestration/v1/stacks/requests.go:
```diff
@@ -265,42 +265,65 @@ type UpdateOptsBuilder interface {
-	for k, v := range opts.TemplateOpts.Files {
-		files[k] = v
+
+	// Template is only required on PUT update
```
We can nix this comment now.

jtopjian commented on openstack/orchestration/v1/stacks/requests.go:
```diff
@@ -340,6 +363,18 @@ func Update(c *gophercloud.ServiceClient, stackName, stackID string, opts Update
 }
 
+// Update accepts an UpdateOpts struct and updates an existing stack using the
+//  http PATCH verb with the values provided. opts.TemplateOpts is required.
```
opts.TemplateOpts is _not_ required?

jtopjian commented on openstack/orchestration/v1/stacks/requests.go:
```diff
@@ -265,42 +265,65 @@ type UpdateOptsBuilder interface {
 	ToStackUpdateMap() (map[string]interface{}, error)
 }
 
+type UpdatePatchOptsBuilder interface {
```
Let's add a doc comment here:

```
// UpdatePatchOptsBuilder is the interface options structs have to satisfy in order
// to be used in the UpdatePatch operation in this package.
```

scott-chaney commented:
@jtopjian Thanks for the feedback! I hope that I addressed all of the points that you mentioned.

> I don't use Gophercloud+Heat, so I haven't tried this first-hand -- have you?
Yes, the info in the documentation notes were gathered from functionally running the changes in GopherCloud in one of our Openstack environments for both `Update `and `UpdatePatch` and then inspecting the stack and the Heat database directly.  I also compared the behavior to the direct Heat API calls.

Let me know if there are any other changes you think would improve this.  Thanks again for the help.
